keylime/ima/ima.py

```diff
@@ -477,7 +498,54 @@ def read_allowlist(al_path=None, checksum="", gpg_sig_file=None, gpg_key_file=No
                 f"Checksum of allowlist does not match! Expected {checksum}, Calculated {calculated_checksum}"
             )
 
+    # Return artifacts as a JSON bundle, ready for sending
+    bundle = {
+        "ima_policy": base64.b64encode(alist_bytes).decode(),
+        "key": base64.b64encode(al_key).decode(),
+        "sig": base64.b64encode(al_sig).decode(),
+        "checksum": checksum,
+    }
+
+    return bundle
+
+
+class SignatureValidationError(Exception):
+    def __init__(self, message, code):
+        self.message = message
+        self.code = code
+        super().__init__(self.message)
+
+
+# Reads allowlist bundles sent to the verifier, validates them, and processes them.
+def unbundle_allowlist(allowlist_bundle, force_verification=False):
+    logger.info("allowlist bundle: %s", allowlist_bundle)
+    allowlist_raw = base64.b64decode(allowlist_bundle["ima_policy"])
+
+    # Verify allowlist signatures if set to enforce
+    if force_verification or config.getboolean("cloud_verifier", "enforce_allow_list_signatures", fallback=False):
+        if not allowlist_bundle.get("sig"):
+            raise SignatureValidationError(
+                message="Verifier is enforcing signature validation, but no signature was provided.", code=405
+            )
+        if not allowlist_bundle.get("key"):
+            raise SignatureValidationError(
+                message="Verifier is enforcing signature validation, but no public key was provided.", code=405
+            )
+
+        allowlist_sig = base64.b64decode(allowlist_bundle.get("sig"))
+        allowlist_sig_key = base64.b64decode(allowlist_bundle.get("key"))
+
+        if not signing.verify_signature(allowlist_sig_key, allowlist_sig, allowlist_raw):
+            raise SignatureValidationError(message="Signature verification for allowlist failed!", code=401)
+        logger.info("Allowlist signature validation passed.")
+
+    return format_allowlist(allowlist_raw, allowlist_bundle["checksum"])
+
+
+# Manipulates allowlists into database-ready format.
+def format_allowlist(alist_bytes, checksum=""):
```

Comment:
Instead of `format_allowlist()` maybe something like `canonicalize_allowlist()`. The "format" is a bit too generic. It's about making sure they are all consistent when serializing into the DB so something like "canonicalize" is a better choice IMO.

Reply:
Agreed. We have a lot of these functions - `update_allowlist`, `process_allowlist`, `format_allowlist`...

Reply:
I've changed this in my latest push.